src/system/mod.rs

```diff
@@ -254,9 +257,9 @@ pub fn chown<S: AsRef<CStr>>(
 pub struct User {
     pub uid: UserId,
     pub gid: GroupId,
-    pub name: String,
+    pub name: SudoString,
     pub gecos: String,
-    pub home: PathBuf,
+    pub home: SudoPath,
```

Comment:
I think this makes sudo-rs more restrictive. this struct comes from the `passwd` database and there's no constraint there that says that HOME has to be UTF-8 encoded so this change will reject users whose home directories are not UTF-8 encoded ... not sure how much of a problem that is in practice

this change was done because `expand_tilde_in_path` operates on a `SudoPath` (extracted from the UTF-8 encoded sudoers file) of the form `~/a/b/c` and it's supposed to replace the `~` with the target (?) user's home directory. if the home directory is not UTF-8 encoded then the expansion will also not be UTF-8. maybe we can make `expand_tilde_in_path` return a regular `PathBuf`? :thinking:

Reply:
well there's the argument that even if we support non utf-8 users, you cannot refer to them directly in the sudoers file. 

If `expand_tilde_in_path` were to return a `PathBuf` I think it would be OK as we only pass this value to `std::env::set_current_dir` afaik.

Reply:
> well there's the argument that even if we support non utf-8 users, you cannot refer to them directly in the sudoers file.

I think this change is still stricter even for users with utf-8 names. the home is retrieved through `libc::getpwnam_r`; it's not something that'll be extracted from the sudoers file (which must be UTF-8 encoded in our implementation). a user could have a UTF-8 name but a non-UTF-8 home directory -- the home directory does not necessarily have to be `/home/user_name` which would certainly be UTF-8 encoded.

> If expand_tilde_in_path were to return a PathBuf I think it would be OK

actually, I think that won't work because of this code in `src/sudo/pipeline.rs`
``` rust
        // expand tildes in the path with the users home directory
        if let Some(dir) = context.chdir.take() {
            context.chdir = Some(dir.expand_tilde_in_path(&context.target_user.name)?)
        }
```
`context.chdir` gets overridden with the return value of `expand_tilde_in_path` so the return type of that function needs to be `SudoPath`

Reply:
damn...